src/mpi_t/qmpi_register.c

```diff
@@ -0,0 +1,402 @@
+        MPIR_QMPI_first_fn_ptrs[i] = NULL;
+
+        for (int j = 1; j <= MPIR_QMPI_num_tools; j++) {
+            if (MPIR_QMPI_pointers[j * MPI_LAST_FUNC_T + i] != NULL) {
+                MPIR_QMPI_first_fn_ptrs[i] = MPIR_QMPI_pointers[j * MPI_LAST_FUNC_T + i];
+                MPIR_QMPI_first_tool_ids[i] = j;
+                break;
+            }
+        }
+
+        if (MPIR_QMPI_first_fn_ptrs[i] == NULL) {
+            MPIR_QMPI_first_fn_ptrs[i] = MPIR_QMPI_pointers[i];
+            MPIR_QMPI_first_tool_ids[i] = 0;
+        }
+    }
+    return MPI_SUCCESS;
+}
+
+int MPII_qmpi_setup(void)
+{
+    int mpi_errno = MPI_SUCCESS;
+    size_t len = 0;
+
+    static bool initialized = false;
+
+    if (!initialized) {
+        initialized = true;
+
+        mpi_errno = MPIR_T_env_init();
+        MPIR_ERR_CHECK(mpi_errno);
+
+        /* Parse environment variable to get the number and list of tools */
+        if (MPIR_CVAR_QMPI_TOOL_LIST == NULL) {
+            MPIR_QMPI_num_tools = 0;
+            goto fn_exit;
+        } else {
+            MPIR_QMPI_num_tools = 1;
+            len = strlen(MPIR_CVAR_QMPI_TOOL_LIST);
+        }
+
+        for (int i = 0; i < len; i++) {
+            if (MPIR_CVAR_QMPI_TOOL_LIST[i] == ':') {
+                MPIR_QMPI_num_tools++;
+            }
+        }
+        MPIR_QMPI_tool_names = MPL_calloc(MPIR_QMPI_num_tools + 1, sizeof(char *), MPL_MEM_OTHER);
+        if (!MPIR_QMPI_tool_names) {
+            MPIR_Err_create_code(mpi_errno, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
+                                 MPI_ERR_OTHER, "**nomem", "**nomem %s", "MPIR_QMPI_num_tools");
+            goto fn_exit;
+        }
+        for (int i = 0; i < MPIR_QMPI_num_tools + 1; i++) {
+            MPIR_QMPI_tool_names[i] =
+                MPL_calloc(QMPI_MAX_TOOL_NAME_LENGTH, sizeof(char), MPL_MEM_OTHER);
+            if (!MPIR_QMPI_tool_names[i]) {
+                MPIR_Err_create_code(mpi_errno, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
+                                     MPI_ERR_OTHER, "**nomem", "**nomem %s", "MPIR_QMPI_num_tools");
+                goto fn_exit;
+            }
+        }
+
+        char *save_ptr, *val;
+        char *tmp_str = MPL_strdup(MPIR_CVAR_QMPI_TOOL_LIST);
+        val = strtok_r(tmp_str, ":", &save_ptr);
+        int tool_num = 1;
+        while (val != NULL) {
+            strncpy(MPIR_QMPI_tool_names[tool_num], val, QMPI_MAX_TOOL_NAME_LENGTH);
+            /* Make sure the string is null terminated */
+            MPIR_QMPI_tool_names[tool_num][QMPI_MAX_TOOL_NAME_LENGTH - 1] = '\0';
+            tool_num++;
+            val = strtok_r(NULL, ":", &save_ptr);
+        }
+        MPL_free(tmp_str);
+        MPIR_QMPI_num_tools = tool_num - 1;
+
+        /* Allocate space for as many tools as we say we support */
+        MPIR_QMPI_pointers = MPL_calloc(1, sizeof(generic_mpi_func *) * MPI_LAST_FUNC_T *
+                                        (MPIR_QMPI_num_tools + 1), MPL_MEM_OTHER);
```

Comment:
The first thing I want to understand is the dimension of all these arrays. Knowing the dimension pretty much enables me to guess their purposes. However, current, this is a bit more difficult to glance. How about use two additional local variables -- `n_funcs` and `n_tools`?

Reply:
Fair enough. I replaced the use of `MPIR_QMPI_num_tools` here with `tool_num` since it already existed. I also added a comment to the declaration of that variable to make it more clear what's going on.

I also replaced `MPI_LAST_FUNC_T` with `num_funcs`.